IdentityCore/src/cache/ios/MSIDKeychainTokenCache.m

```diff
@@ -22,30 +22,186 @@
+}
+
++ (MSIDKeychainTokenCache *)defaultKeychainCache
 {
-    return NO;
+    if (!_defaultCache)
+    {
+        _defaultCache = [[MSIDKeychainTokenCache alloc] initWithGroup:MSIDKeychainTokenCache.adalAccessGroup];
+    }
+    
+    return _defaultCache;
+}
+
++ (void)setDefaultKeychainCache:(MSIDKeychainTokenCache *)defaultKeychainCache
+{
+    if (!defaultKeychainCache)
+    {
+        return;
+    }
+    
+    _defaultCache = defaultKeychainCache;
+}
+
+- (instancetype)initWithGroup:(NSString *)accessGroup
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    if (!accessGroup)
+    {
+        MSID_LOG_INFO(nil, @"Keychain initialization with nil accessGroup.");
+        return nil;
+    }
+    
+    MSID_LOG_INFO(nil, @"Using keychain accessGroup: %@", _PII_NULLIFY(accessGroup));
+    MSID_LOG_INFO_PII(nil, @"Using keychain accessGroup: %@", accessGroup);
+    
+    _accessGroup = accessGroup;
+    
+    return self;
+}
+
+#pragma mark - MSIDTokenCacheDataSource
+
+- (BOOL)setItem:(MSIDToken *)item
+        withKey:(MSIDTokenCacheKey *)key
+     serializer:(id<MSIDTokenSerializer>)serializer
+        context:(id<MSIDRequestContext>)context
+          error:(NSError **)error
+{
+    assert(key);
+    assert(serializer);
+    
+    MSID_LOG_INFO(context, @"Set keychain item, key info (account: %@ service: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service));
+    MSID_LOG_INFO_PII(context, @"Set keychain item, key info (account: %@ service: %@)", key.account, key.service);
+    MSID_LOG_INFO_PII(context, @"Item info %@", item);
+    
+    if (!key.service || !key.account)
+    {
+        if (error)
+        {
+            // TODO: Use proper domain & error code.
+            *error = [[NSError alloc] initWithDomain:@"MSIDErrorDomain" code:-1 userInfo:@{NSLocalizedDescriptionKey:@"Key is not valid. Make sure service and account are not nil."}];
+        }
+        MSID_LOG_ERROR(context, @"Set keychain item with invalid key.");
+        return NO;
+    }
+
+    NSData *itemData = [serializer serialize:item];
+    if (!itemData)
+    {
+        if (error)
+        {
+            *error = [[NSError alloc] initWithDomain:@"MSIDErrorDomain" code:-1 userInfo:@{NSLocalizedDescriptionKey:@"Failed to serialize token item."}];
+        }
+        MSID_LOG_ERROR(context, @"Failed to serialize token item.");
+        return NO;
+    }
+    
+    NSMutableDictionary *query = [@{(id)kSecClass : (id)kSecClassGenericPassword} mutableCopy];
+    [query setObject:key.service forKey:(id)kSecAttrService];
+    [query setObject:key.account forKey:(id)kSecAttrAccount];
+    [query setObject:self.accessGroup forKey:(id)kSecAttrAccessGroup];
+    // Backward compatibility with ADAL.
+    [query setObject:[s_libraryString dataUsingEncoding:NSUTF8StringEncoding] forKey:(id)kSecAttrGeneric];
+    // Backward compatibility with MSAL.
+    [query setObject:[NSNumber numberWithUnsignedInt:item.tokenType] forKey:(id)kSecAttrType];
```

Comment:
Older ADAL version don't set this attribute, so this will break coexistence. If we want to keep it, we should set it only for MSAL tokens.

Reply:
I don't think it will break compatibility with ADAL. In ADAL the query will not contain kSecAttrType, so all items that contain this key should be available in ADAL as well. In MSID we put this attribute only in "set item" method, the query logic is identical to ADAL.